File: backend/src/routes/messages.ts
import { Router, Response } from 'express';
import { body, validationResult } from 'express-validator';
import prisma from '../config/database';
import { AuthRequest } from '../types';
import { authenticateToken, requireAdmin } from '../middleware/auth';
import { emitRealtimeEvent } from '../utils/eventStore';

const router = Router();

// Get contact messages (Admin only)
router.get('/contact', [authenticateToken, requireAdmin], async (req: AuthRequest, res: Response) => {
  try {
    const messages = await prisma.contactMessage.findMany({
      include: {
        center: {
          select: { id: true, name: true }
        },
        sender: {
          select: { id: true, name: true, email: true }
        }
      },
      orderBy: { createdAt: 'desc' }
    });

    const transformedMessages = messages.map(message => ({
      id: message.id,
      centerName: message.center.name,
      centerId: message.centerId,
      senderName: message.senderName,
      senderEmail: message.senderEmail,
      subject: message.subject,
      message: message.message,
      inquiryType: message.inquiryType,
      timestamp: message.createdAt,
      status: message.status.toLowerCase()
    }));

    res.json({ messages: transformedMessages });
  } catch (error) {
    console.error('Get contact messages error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Send contact message
router.post('/contact', [
  authenticateToken,
  body('centerId').isString(),
  body('subject').trim().isLength({ min: 5 }),
  body('message').trim().isLength({ min: 10 }),
  body('inquiryType').isString()
], async (req: AuthRequest, res: Response) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { centerId, subject, message, inquiryType } = req.body;

    // Verify center exists
    const center = await prisma.communityCenter.findUnique({
      where: { id: centerId }
    });

    if (!center) {
      return res.status(404).json({ error: 'Center not found' });
    }

    const contactMessage = await prisma.contactMessage.create({
      data: {
        centerId,
        senderUserId: req.user!.id,
        senderName: req.user!.name,
        senderEmail: req.user!.email,
        subject,
        message,
        inquiryType,
        status: 'PENDING'
      }
    });

    // Emit real-time event for admin notifications
    emitRealtimeEvent('new-contact-message', {
      id: contactMessage.id,
      centerName: center.name,
      centerId: contactMessage.centerId,
      senderName: contactMessage.senderName,
      senderEmail: contactMessage.senderEmail,
      subject: contactMessage.subject,
      inquiryType: contactMessage.inquiryType,
      timestamp: contactMessage.createdAt
    });

    res.status(201).json({
      message: 'Contact message sent successfully',
      contactMessage: {
        id: contactMessage.id,
        centerName: center.name,
        centerId: contactMessage.centerId,
        senderName: contactMessage.senderName,
        senderEmail: contactMessage.senderEmail,
        subject: contactMessage.subject,
        message: contactMessage.message,
        inquiryType: contactMessage.inquiryType,
        timestamp: contactMessage.createdAt,
        status: contactMessage.status.toLowerCase()
      }
    });
  } catch (error) {
    console.error('Send contact message error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get message threads for a center
router.get('/threads/:centerId', [authenticateToken], async (req: AuthRequest, res: Response) => {
  try {
    const { centerId } = req.params;

    // Verify user has access to this center (admin or center manager)
    if (req.user!.role !== 'ADMIN') {
      const center = await prisma.communityCenter.findFirst({
        where: {
          id: centerId,
          managerId: req.user!.id
        }
      });

      if (!center) {
        return res.status(403).json({ error: 'Access denied to this center' });
      }
    }

    const threads = await prisma.messageThread.findMany({
      where: {
        participants: {
          some: { id: centerId }
        }
      },
      include: {
        participants: {
          select: { id: true, name: true }
        },
        messages: {
          orderBy: { createdAt: 'desc' },
          take: 1,
          include: {
            sender: {
              select: { id: true, name: true }
            }
          }
        }
      },
      orderBy: { lastActivity: 'desc' }
    });

    const transformedThreads = threads.map(thread => ({
      id: thread.id,
      participants: thread.participants.map(p => p.id),
      participantNames: thread.participants.map(p => p.name),
      subject: thread.subject,
      lastMessage: thread.messages[0] ? {
        id: thread.messages[0].id,
        threadId: thread.messages[0].threadId,
        senderId: thread.messages[0].senderId,
        senderName: thread.messages[0].sender.name,
        content: thread.messages[0].content,
        timestamp: thread.messages[0].createdAt,
        read: thread.messages[0].read
      } : undefined,
      lastActivity: thread.lastActivity,
      messageCount: thread.messageCount
    }));

    res.json({ threads: transformedThreads });
  } catch (error) {
    console.error('Get message threads error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get messages in a thread
router.get('/threads/:threadId/messages', [authenticateToken], async (req: AuthRequest, res: Response) => {
  try {
    const { threadId } = req.params;

    // Verify user has access to this thread
    const thread = await prisma.messageThread.findFirst({
      where: {
        id: threadId,
        participants: {
          some: {
            OR: [
              { managerId: req.user!.id },
              ...(req.user!.role === 'ADMIN' ? [{}] : [])
            ]
          }
        }
      },
      include: {
        participants: {
          select: { id: true, name: true }
        }
      }
    });

    if (!thread) {
      return res.status(404).json({ error: 'Thread not found or access denied' });
    }

    const messages = await prisma.centerMessage.findMany({
      where: { threadId },
      include: {
        sender: {
          select: { id: true, name: true }
        }
      },
      orderBy: { createdAt: 'asc' }
    });

    const transformedMessages = messages.map(message => ({
      id: message.id,
      threadId: message.threadId,
      senderId: message.senderId,
      senderName: message.sender.name,
      content: message.content,
      timestamp: message.createdAt,
      read: message.read
    }));

    res.json({ messages: transformedMessages });
  } catch (error) {
    console.error('Get thread messages error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Send message to thread
router.post('/threads/:threadId/messages', [
  authenticateToken,
  body('content').trim().isLength({ min: 1 })
], async (req: AuthRequest, res: Response) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { threadId } = req.params;
    const { content } = req.body;

    // Find a center managed by this user to send from
    const senderCenter = await prisma.communityCenter.findFirst({
      where: {
        OR: [
          { managerId: req.user!.id },
          ...(req.user!.role === 'ADMIN' ? [{ verified: true }] : [])
        ]
      }
    });

    if (!senderCenter) {
      return res.status(403).json({ error: 'No center found to send message from' });
    }

    // Verify thread exists and user has access
    const thread = await prisma.messageThread.findFirst({
      where: {
        id: threadId,
        participants: {
          some: { id: senderCenter.id }
        }
      }
    });

    if (!thread) {
      return res.status(404).json({ error: 'Thread not found or access denied' });
    }

    const message = await prisma.centerMessage.create({
      data: {
        threadId,
        senderId: senderCenter.id,
        content,
        read: false
      }
    });

    // Update thread activity and message count
    await prisma.messageThread.update({
      where: { id: threadId },
      data: {
        lastActivity: new Date(),
        messageCount: { increment: 1 }
      }
    });

    // Emit real-time event for new message
    emitRealtimeEvent('new-message', {
      id: message.id,
      threadId: message.threadId,
      senderId: message.senderId,
      senderName: senderCenter.name,
      content: message.content,
      timestamp: message.createdAt,
      read: message.read
    });

    res.status(201).json({
      message: 'Message sent successfully',
      centerMessage: {
        id: message.id,
        threadId: message.threadId,
        senderId: message.senderId,
        senderName: senderCenter.name,
        content: message.content,
        timestamp: message.createdAt,
        read: message.read
      }
    });
  } catch (error) {
    console.error('Send message error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Create new message thread
router.post('/threads', [
  authenticateToken,
  body('participantIds').isArray().isLength({ min: 2 }),
  body('subject').trim().isLength({ min: 5 }),
  body('initialMessage').trim().isLength({ min: 1 })
], async (req: AuthRequest, res: Response) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { participantIds, subject, initialMessage } = req.body;

    // Find sender center
    const senderCenter = await prisma.communityCenter.findFirst({
      where: {
        OR: [
          { managerId: req.user!.id },
          ...(req.user!.role === 'ADMIN' ? [{ verified: true }] : [])
        ]
      }
    });

    if (!senderCenter) {
      return res.status(403).json({ error: 'No center found to send message from' });
    }

    // Verify all participants are verified centers
    const participants = await prisma.communityCenter.findMany({
      where: {
        id: { in: participantIds },
        verified: true
      }
    });

    if (participants.length !== participantIds.length) {
      return res.status(400).json({ error: 'Some participants are not verified centers' });
    }

    // Create thread
    const thread = await prisma.messageThread.create({
      data: {
        subject,
        lastActivity: new Date(),
        messageCount: 1,
        participants: {
          connect: participantIds.map((id: string) => ({ id }))
        }
      }
    });

    // Create initial message
    const message = await prisma.centerMessage.create({
      data: {
        threadId: thread.id,
        senderId: senderCenter.id,
        content: initialMessage,
        read: false
      }
    });

    res.status(201).json({
      message: 'Thread created successfully',
      thread: {
        id: thread.id,
        participants: participantIds,
        participantNames: participants.map(p => p.name),
        subject: thread.subject,
        lastMessage: {
          id: message.id,
          threadId: message.threadId,
          senderId: message.senderId,
          senderName: senderCenter.name,
          content: message.content,
          timestamp: message.createdAt,
          read: message.read
        },
        lastActivity: thread.lastActivity,
        messageCount: thread.messageCount
      }
    });
  } catch (error) {
    console.error('Create thread error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: go-backend/internal/http/handlers/messages.go
package handlers

import (
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"communitycentresplatform/go-backend/internal/ctxutil"
	"communitycentresplatform/go-backend/internal/db"
)

// GET /api/messages/contact - Get contact messages (Admin only)
func GetContactMessages(c *gin.Context) {
	gdb := ctxutil.DBFrom(c)
	if gdb == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "db unavailable"})
		return
	}

	messages, err := db.ListContactMessages(gdb, "", 0, 0)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch contact messages"})
		return
	}

	// Transform to match Node.js format
	transformed := make([]gin.H, len(messages))
	for i, msg := range messages {
		transformed[i] = gin.H{
			"id":          msg.ID,
			"centerName":  msg.Center.Name,
			"centerId":    msg.CenterID,
			"senderName":  msg.SenderName,
			"senderEmail": msg.SenderEmail,
			"subject":     msg.Subject,
			"message":     msg.Message,
			"inquiryType": msg.InquiryType,
			"timestamp":   msg.CreatedAt,
			"status":      strings.ToLower(string(msg.Status)),
		}
	}

	c.JSON(http.StatusOK, gin.H{"messages": transformed})
}

// POST /api/messages/contact - Send contact message
func SendContactMessage(c *gin.Context) {
	gdb := ctxutil.DBFrom(c)
	if gdb == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "db unavailable"})
		return
	}

	var req struct {
		CenterID    string `json:"centerId" binding:"required"`
		Subject     string `json:"subject" binding:"required,min=5"`
		Message     string `json:"message" binding:"required,min=10"`
		InquiryType string `json:"inquiryType" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}

	userID := ctxutil.UserIDFrom(c)
	userName := ctxutil.NameFrom(c)
	userEmail := ctxutil.EmailFrom(c)

	// Parse center ID
	centerID, err := uuid.Parse(req.CenterID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid centerId"})
		return
	}

	// Verify center exists
	center, err := db.FindCenterByID(gdb, centerID)
	if err != nil || center == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "center not found"})
		return
	}

	// Parse user ID
	senderUserID, _ := uuid.Parse(userID)

	// Create contact message
	contactMsg := &db.ContactMessage{
		CenterID:     centerID,
		SenderUserID: senderUserID,
		SenderName:   userName,
		SenderEmail:  userEmail,
		Subject:      req.Subject,
		Message:      req.Message,
		InquiryType:  req.InquiryType,
		Status:       db.ContactStatusPending,
	}

	if err := db.CreateContactMessage(gdb, contactMsg); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to send contact message"})
		return
	}

	// Emit real-time event for admin notifications
	if br := ctxutil.BrokerFrom(c); br != nil {
		br.EmitNewMessage("admin", gin.H{
			"type":        "new-contact-message",
			"id":          contactMsg.ID,
			"centerName":  center.Name,
			"centerId":    contactMsg.CenterID,
			"senderName":  contactMsg.SenderName,
			"senderEmail": contactMsg.SenderEmail,
			"subject":     contactMsg.Subject,
			"inquiryType": contactMsg.InquiryType,
			"timestamp":   contactMsg.CreatedAt,
		})
	}

	c.JSON(http.StatusCreated, gin.H{
		"message": "Contact message sent successfully",
		"contactMessage": gin.H{
			"id":          contactMsg.ID,
			"centerName":  center.Name,
			"centerId":    contactMsg.CenterID,
			"senderName":  contactMsg.SenderName,
			"senderEmail": contactMsg.SenderEmail,
			"subject":     contactMsg.Subject,
			"message":     contactMsg.Message,
			"inquiryType": contactMsg.InquiryType,
			"timestamp":   contactMsg.CreatedAt,
			"status":      strings.ToLower(string(contactMsg.Status)),
		},
	})
}

// GET /api/messages/threads/:centerId - Get message threads for a center
func GetThreadsForCenter(c *gin.Context) {
	gdb := ctxutil.DBFrom(c)
	if gdb == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "db unavailable"})
		return
	}

	centerIDStr := c.Param("centerId")
	centerID, err := uuid.Parse(centerIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid center id"})
		return
	}

	userID := ctxutil.UserIDFrom(c)
	role := ctxutil.RoleFrom(c)

	// Verify user has access (admin or center manager)
	if role != "ADMIN" {
		uid, _ := uuid.Parse(userID)
		center, _ := db.FindCenterByID(gdb, centerID)
		if center == nil || center.ManagerID == nil || *center.ManagerID != uid {
			c.JSON(http.StatusForbidden, gin.H{"error": "access denied to this center"})
			return
		}
	}

	// Get threads
	threads, err := db.ListThreadsForCenter(gdb, centerID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch threads"})
		return
	}

	// Transform to match Node.js format
	transformed := make([]gin.H, len(threads))
	for i, thread := range threads {
		// Get participant IDs and names
		participantIDs := make([]uuid.UUID, len(thread.Participants))
		participantNames := make([]string, len(thread.Participants))
		for j, p := range thread.Participants {
			participantIDs[j] = p.ID
			participantNames[j] = p.Name
		}

		// Get last message
		var lastMessage interface{}
		if len(thread.Messages) > 0 {
			msg := thread.Messages[0]
			lastMessage = gin.H{
				"id":         msg.ID,
				"threadId":   msg.ThreadID,
				"senderId":   msg.SenderID,
				"senderName": msg.Sender.Name,
				"content":    msg.Content,
				"timestamp":  msg.CreatedAt,
				"read":       msg.Read,
			}
		}

		transformed[i] = gin.H{
			"id":               thread.ID,
			"participants":     participantIDs,
			"participantNames": participantNames,
			"subject":          thread.Subject,
			"lastMessage":      lastMessage,
			"lastActivity":     thread.LastActivity,
			"messageCount":     thread.MessageCount,
		}
	}

	c.JSON(http.StatusOK, gin.H{"threads": transformed})
}

// GET /api/messages/threads/:threadId/messages - Get messages in a thread
func GetThreadMessages(c *gin.Context) {
	gdb := ctxutil.DBFrom(c)
	if gdb == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "db unavailable"})
		return
	}

	threadIDStr := c.Param("threadId")
	threadID, err := uuid.Parse(threadIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid thread id"})
		return
	}

	userID := ctxutil.UserIDFrom(c)
	role := ctxutil.RoleFrom(c)

	// Verify thread exists
	thread, err := db.FindThreadByID(gdb, threadID)
	if err != nil || thread == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "thread not found or access denied"})
		return
	}

	// Verify access (admin or participant's manager)
	hasAccess := role == "ADMIN"
	if !hasAccess {
		uid, _ := uuid.Parse(userID)
		for _, participant := range thread.Participants {
			if participant.ManagerID != nil && *participant.ManagerID == uid {
				hasAccess = true
				break
			}
		}
	}

	if !hasAccess {
		c.JSON(http.StatusForbidden, gin.H{"error": "thread not found or access denied"})
		return
	}

	// Get messages
	messages, err := db.ListMessagesInThread(gdb, threadID, 0, 0)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch messages"})
		return
	}

	// Transform to match Node.js format
	transformed := make([]gin.H, len(messages))
	for i, msg := range messages {
		transformed[i] = gin.H{
			"id":         msg.ID,
			"threadId":   msg.ThreadID,
			"senderId":   msg.SenderID,
			"senderName": msg.Sender.Name,
			"content":    msg.Content,
			"timestamp":  msg.CreatedAt,
			"read":       msg.Read,
		}
	}

	c.JSON(http.StatusOK, gin.H{"messages": transformed})
}

// POST /api/messages/threads/:threadId/messages - Send message to thread
func SendThreadMessage(c *gin.Context) {
	gdb := ctxutil.DBFrom(c)
	if gdb == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "db unavailable"})
		return
	}

	threadIDStr := c.Param("threadId")
	threadID, err := uuid.Parse(threadIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid thread id"})
		return
	}

	var req struct {
		Content string `json:"content" binding:"required,min=1"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}

	userID := ctxutil.UserIDFrom(c)
	role := ctxutil.RoleFrom(c)
	uid, _ := uuid.Parse(userID)

	// Find a center managed by this user to send from
	var senderCenter *db.CommunityCenter
	if role == "ADMIN" {
		// Admin can send from any verified center
		centers, _ := db.ListCenters(gdb, db.CenterFilters{VerificationStatus: "verified", Limit: 1})
		if len(centers) > 0 {
			senderCenter = &centers[0]
		}
	} else {
		// Find center managed by this user
		centers, _ := db.ListCenters(gdb, db.CenterFilters{Limit: 100})
		for _, center := range centers {
			if center.ManagerID != nil && *center.ManagerID == uid {
				senderCenter = &center
				break
			}
		}
	}

	if senderCenter == nil {
		c.JSON(http.StatusForbidden, gin.H{"error": "no center found to send message from"})
		return
	}

	// Verify thread exists and user has access
	thread, err := db.FindThreadByID(gdb, threadID)
	if err != nil || thread == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "thread not found or access denied"})
		return
	}

	// Check if sender's center is a participant
	isParticipant := false
	for _, participant := range thread.Participants {
		if participant.ID == senderCenter.ID {
			isParticipant = true
			break
		}
	}

	if !isParticipant {
		c.JSON(http.StatusForbidden, gin.H{"error": "thread not found or access denied"})
		return
	}

	// Create message (transaction updates thread metadata)
	message, err := db.CreateMessage(gdb, threadID, senderCenter.ID, req.Content)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to send message"})
		return
	}

	// Emit real-time event
	if br := ctxutil.BrokerFrom(c); br != nil {
		br.EmitNewMessage(threadID.String(), gin.H{
			"id":         message.ID,
			"threadId":   message.ThreadID,
			"senderId":   message.SenderID,
			"senderName": senderCenter.Name,
			"content":    message.Content,
			"timestamp":  message.CreatedAt,
			"read":       message.Read,
		})
	}

	c.JSON(http.StatusCreated, gin.H{
		"message": "Message sent successfully",
		"centerMessage": gin.H{
			"id":         message.ID,
			"threadId":   message.ThreadID,
			"senderId":   message.SenderID,
			"senderName": senderCenter.Name,
			"content":    message.Content,
			"timestamp":  message.CreatedAt,
			"read":       message.Read,
		},
	})
}

// POST /api/messages/threads - Create new message thread
func CreateThread(c *gin.Context) {
	gdb := ctxutil.DBFrom(c)
	if gdb == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "db unavailable"})
		return
	}

	var req struct {
		ParticipantIDs []string `json:"participantIds" binding:"required,min=2"`
		Subject        string   `json:"subject" binding:"required,min=5"`
		InitialMessage string   `json:"initialMessage" binding:"required,min=1"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}

	userID := ctxutil.UserIDFrom(c)
	role := ctxutil.RoleFrom(c)
	uid, _ := uuid.Parse(userID)

	// Find sender center
	var senderCenter *db.CommunityCenter
	if role == "ADMIN" {
		centers, _ := db.ListCenters(gdb, db.CenterFilters{VerificationStatus: "verified", Limit: 1})
		if len(centers) > 0 {
			senderCenter = &centers[0]
		}
	} else {
		centers, _ := db.ListCenters(gdb, db.CenterFilters{Limit: 100})
		for _, center := range centers {
			if center.ManagerID != nil && *center.ManagerID == uid {
				senderCenter = &center
				break
			}
		}
	}

	if senderCenter == nil {
		c.JSON(http.StatusForbidden, gin.H{"error": "no center found to send message from"})
		return
	}

	// Parse participant IDs
	participantIDs := make([]uuid.UUID, len(req.ParticipantIDs))
	for i, idStr := range req.ParticipantIDs {
		id, err := uuid.Parse(idStr)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid participantIds"})
			return
		}
		participantIDs[i] = id
	}

	// Create thread with participants
	thread, err := db.CreateMessageThread(gdb, req.Subject, participantIDs)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to create thread"})
		return
	}

	// Create initial message
	message, err := db.CreateMessage(gdb, thread.ID, senderCenter.ID, req.InitialMessage)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to send initial message"})
		return
	}

	// Get participant names
	participantNames := make([]string, len(thread.Participants))
	for i, p := range thread.Participants {
		participantNames[i] = p.Name
	}

	c.JSON(http.StatusCreated, gin.H{
		"message": "Thread created successfully",
		"thread": gin.H{
			"id":               thread.ID,
			"participants":     participantIDs,
			"participantNames": participantNames,
			"subject":          thread.Subject,
			"lastMessage": gin.H{
				"id":         message.ID,
				"threadId":   message.ThreadID,
				"senderId":   message.SenderID,
				"senderName": senderCenter.Name,
				"content":    message.Content,
				"timestamp":  message.CreatedAt,
				"read":       message.Read,
			},
			"lastActivity": thread.LastActivity,
			"messageCount": thread.MessageCount,
		},
	})
}
